src/AspNet.Security.OAuth.Strava/StravaAuthenticationHandler.cs

```diff
@@ -32,6 +34,19 @@ public class StravaAuthenticationHandler : OAuthHandler<StravaAuthenticationOpti
         {
         }
 
+        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
+        {
+            var scopeValues = Request.Query[OAuthChallengeProperties.ScopeKey];
+
+            if (scopeValues.Count == 0 || scopeValues.First().Split(",").Length < Options.Scope.Count)
+            {
+                var scopeEx = new Exception("User did not grant access to requested scopes.");
+                return HandleRequestResult.Fail(scopeEx);
+            }
+
+            return await base.HandleRemoteAuthenticateAsync();
+        }
+
```

Comment:
What about the suggested approach below instead?

Here any HTTP or standard OAuth authentication/authorization errors are handled first, then we give the user the _option_ of handling access being denied (as they effectively denied permission for a requested scope) in a custom manner of their own using the existing `AccessDenied` event.

Then you can just hook up a `OAuthEvents.AccessDenied` handler and the code will follow this flow:

https://github.com/dotnet/aspnetcore/blob/ef5ab43b6f7bfeaa3cfb2c0735715a7cb69fda37/src/Security/Authentication/Core/src/RemoteAuthenticationHandler.cs#L244-L285

In your case, you can throw an exception, where as another user may chose to just continue regardless as they may consider the particular scope denied to be optional, rather than block the entire authentication flow due to _something_ being missing.

It also has the benefit of redirecting the user to the `AccessDeniedPath` for custom messaging and/or logic, rather than to an HTTP 500 error due to the exception.

```suggestion
        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
        {
            HandleRequestResult result = await base.HandleRemoteAuthenticateAsync();

            if (result.Succeeded)
            {
                var scopeValues = Request.Query[OAuthChallengeProperties.ScopeKey];

                if (scopeValues.Count == 0 || scopeValues[0].Split(",").Length < Options.Scope.Count)
                {
                    result = await HandleAccessDeniedErrorAsync(result.Properties);
                }
            }

            return result;
        }
```